kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/GCSToBQLoadRunnable.java

```diff
@@ -0,0 +1,232 @@
+  // max number of files we can load in a single load job
+  private static long MAX_LOAD_SIZE_B = 15 * 1000000000000L; // 15TB
+  // max total size (in bytes) of the files we can load in a single load job.
+  private static String SOURCE_URI_FORMAT = "gs://%s/%s";
+  public static final Pattern METADATA_TABLE_PATTERN =
+          Pattern.compile("(?<project>[^:]+):(?<dataset>[^.]+).(?<table>.+)");
+
+  /**
+   * Create a {@link GCSToBQLoadRunnable} with the given bigquery, bucket, and ms wait interval.
+   * @param bigQuery the {@link BigQuery} instance.
+   * @param bucket the the GCS bucket to read from.
+   */
+  public GCSToBQLoadRunnable(BigQuery bigQuery, Bucket bucket) {
+    this.bigQuery = bigQuery;
+    this.bucket = bucket;
+    this.activeJobs = new HashMap<>();
+    this.claimedBlobs = new HashSet<>();
+  }
+
+  private Map<TableId, List<Blob>> getBlobsUpToLimit() {
+    Map<TableId, List<Blob>> tableToURIs = new HashMap<>();
+    //List<String> URIs = new ArrayList<>();
+    Map<TableId, Long> tableToCurrentLoadSize = new HashMap<>();
+
+    Page<Blob> list = bucket.list();
+    for (Blob blob : list.iterateAll()) {
+      TableId table = getTableFromBlob(blob);
+
+      if (table == null || claimedBlobs.contains(blob)) {
+        // don't do anything if:
+        // 1. we don't know what table this should be uploaded to or
+        // 2. this blob is already claimed by a currently-running job.
+        continue;
+      }
+
+      if (!tableToURIs.containsKey(table)) {
+        // initialize maps, if we haven't seen this table before.
+        tableToURIs.put(table, new ArrayList<>());
+        tableToCurrentLoadSize.put(table, 0L);
+      }
+
+      long newSize = tableToCurrentLoadSize.get(table) + blob.getSize();
+      // if this file does not cause us to exceed our per-request quota limits...
+      if (newSize <= MAX_LOAD_SIZE_B && tableToURIs.get(table).size() < FILE_LOAD_LIMIT) {
+        // ...add the file (and update the load size)
+        tableToURIs.get(table).add(blob);
+        tableToCurrentLoadSize.put(table, newSize);
+      }
+    }
+    return tableToURIs;
+  }
+
+  /**
+   * Given a blob, return the {@link TableId} this blob should be inserted into.
+   * @param blob the blob
+   * @return the TableId this data should be loaded into, or null if we could not tell what
+   *         table it should be loaded into.
+   */
+  private TableId getTableFromBlob(Blob blob) {
+    if (blob.getMetadata() == null
+        || blob.getMetadata().get(GCSToBQWriter.GCS_METADATA_TABLE_KEY) == null) {
+      logger.error("Found blob {}\\{} with no metadata.", blob.getBucket(), blob.getName());
+      return null;
+    }
+
+    String serializedTableId = blob.getMetadata().get(GCSToBQWriter.GCS_METADATA_TABLE_KEY);
+    Matcher matcher = METADATA_TABLE_PATTERN.matcher(serializedTableId);
+    String project = matcher.group("project");
+    String dataset = matcher.group("dataset");
+    String table = matcher.group("table");
+
```

Comment:
can you add a debug statement here so we can see what project/dataset/id are for a given blob if we need to